src/plugins/intel_gpu/src/graph/graph_optimizer/mark_shape_of_subgraphs.cpp

```diff
@@ -86,7 +86,13 @@ bool mark_shape_of_subgraphs::can_mark_node(const program_node& node) {
         return false;
     }
 
-    return true;
+    auto available_impl_types = node.type()->get_available_impl_types(node);
+    auto cpu_impl_found = available_impl_types.find(impl_types::cpu) != available_impl_types.end();
```

Comment:
Could you clarify what's the issue exactly? I intentionally removed this code previously, as a node is a part of shape flow regardless having or not having CPU impl type. The decision for preferred impl type is done later when we iterate over available impls list and we just skip GPU impls for nodes on the shape flow.

Reply:
The impl type is not changed despite the primitive has no CPU impl. So, it makes SegFault.

Reply:
@vladimir-paramuzov Where did you expect the temporarily selected CPU impl type to change?

Reply:
@vladimir-paramuzov Did you expect to change this within the layout_optimizer::get_preferred_impl_type() function?

Reply:
@vladimir-paramuzov For example, after a reorder node is set as in_shape_of_subgraph, can it select the other impl? validation function of ImplementationManger will only return false. Am I misunderstanding something?
https://github.com/openvinotoolkit/openvino/blob/65a8f393ba01db7ae40606d29045518fa9d55c74/src/plugins/intel_gpu/src/graph/impls/registry/reorder_impls.cpp#L39

Reply:
Current logic expects that reorder on the shape flow will use CPU impl due to the predicates specified for the impls:

https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_gpu/src/graph/impls/registry/reorder_impls.cpp#L31-L44

As you can see, all GPU impls reject nodes on the shape flow, and CPU impl is allowed only for the shape flow.

Reply:
@vladimir-paramuzov Do you mean the node on the shape flow cannot be executed with ocl impl?

Reply:
> Could you clarify what's the issue exactly? I intentionally removed this code previously, as a node is a part of shape flow regardless having or not having CPU impl type. The decision for preferred impl type is done later when we iterate over available impls list and we just skip GPU impls for nodes on the shape flow.

Please see the below issued graph.
```mermaid
graph TD
    A("A: stridedslice (in shape flow/CPU impl)")
    B("B: stridedslice.out0 (in shape flow/CPU impl)") 
    C("C: convert:reorder_compressed_to_fp16 (in shape flow/CPU impl)")
    D("D: convolution (Not in shape flow/GPU impl)")
    A-->B-->C-->D
```
After graph optimization, a new reorder is added between C and D.
```mermaid
graph TD
    A("A: stridedslice (in shape flow/CPU impl)")
    B("B: stridedslice.out0 (in shape flow/CPU impl)") 
    C("C: convert:reorder_compressed_to_fp16 (in shape flow/CPU impl)")
    D("D: convolution (Not in shape flow/GPU impl)")
    F("F: reorder:fsv16 (Not in shape flow/ANY)")
    A-->B-->C-->F-->D
```
In the first run, when the shape is updated, this added reorder(F) is set to "in shape flow", which causes SegFault because CPU impls do not support fsv16 format.